incl/lib/LevelParser.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+class LevelParser {
+    public $data = [];
+
+    public function __construct($levelString)
+    {
+        $this->data = self::map($levelString, ':');
+    }
+
+    /**
+     * Convert GD's weird "keyed" array strings to dictionary. Works with out-of-order keys.
+     *
+     * @param $list string
+     * @param $separator string
+     * @return array
+     */
+    public static function map($list, $separator) {
+        $bits = explode($separator, $list);
+        $array = [];
+        for ($i = 1; $i < count($bits); $i += 2) {
+            $array[$bits[$i - 1]] = $bits[$i];
+        }
+        return $array;
+    }
+
+    /**
+     * Flatten dictionary to single delimited string
+     *
+     * @param $dict array|object
+     * @param $separator string
+     * @return string
+     */
+    public static function demap($dict, $separator) {
+        $string = '';
+        foreach ($dict as $key => $value) {
+            $string[] .= "${separator}${key}${separator}${value}";
+        }
+        return $string;
+    }
+
+    /**
+     * @param $string string
+     * @return false|string
+     */
+    public static function base64_urlencode($string) {
+        return base64_encode(strtr($string, '+/', '-_'));
+    }
+
+    /**
+     * @param $string string
+     * @return false|string
+     */
+    public static function base64_urldecode($string) {
+        return base64_decode(strtr($string, '-_', '+/'), true);
+    }
+
+    /**
+     * Validates level data for various things (currently only ACE exploit)
+     *
+     * @return bool
+     */
+    public function validate() {
+        // Ocular Miracle ain't happening
+        $max_level_size = 50 * 1024 * 1024; // 50 MB Max uncompressed size
```

Comment:
keep in mind that any level over this size could bypass the check entirely. it wouldn't be impossible to add padding in ways that inflate the size of the level while keeping it working on desktop platforms...